File: java/algorithms/sorting/mergesort.java
public class MergeSort {
    public static void main(String[] args) {
        int[] arr = {12, 11, 13, 5, 6, 7};
        System.out.println("Original array:");
        for (int i : arr){
            System.out.print(i + " ");
        }

        mergeSort(arr, 0, arr.length-1);
        System.out.println("\n Sorted array:");
        for (int i : arr){
            System.out.print(i + " ");
        }
    }

    // Input: an unsorted array A[0..n-1]
    // Input: an integer l, the left-most index
    // Input: an integer r, the right-most index
    // Output: the array A, sorted in non-decreasing order
    public static int[] mergeSort(int[] A, int l, int r) {
        // Check if there is anything to sort
        if (l < r) {

            // Compute the middle index
            int m = (l + r) / 2;

            // Sort the left-hand side
            mergeSort(A, l, m);

            // Sort the right-hand side
            mergeSort(A, m + 1, r);

            // Merge them back together
            merge(A, l, m, r);

        }
        return A;
    }

    // Input: an array A[0..n-1]
    // Input: an integer l, the left-most index
    // Input: an integer m, the middle index
    // Input: an integer r, the right-most index
    public static void merge(int[] A, int l, int m, int r) {
        // Create a temporary array B
        int[] B = new int[r - l + 1];

        int i = 0; // First unused index of B
        int s = l; // Top of the left stack
        int t = m + 1; // Top of the right stack

        // Keep taking the smallest value as long as both
        // stacks still have values
        while (s <= m && t <= r) {
            if (A[s] <= A[t]) {
                // Take from left stack
                B[i] = A[s];
                s++;
            }
            else {
                // Take from right stack
                B[i] = A[t];
                t++;
            }
            i++;
        }

        // Copy the remaining elements from the left or right stack, if any
        while (s <= m) {
            B[i] = A[s];
            s++;
            i++;
        }
        while (t <= r) {
            B[i] = A[t];
            t++;
            i++;
        }

        // Sorting is now complete. Just copy the newly sorted part over the original
        for (i = 0; i < B.length; i++) {
            A[l + i] = B[i];
        }
    }
}

File: cs/algorithms/sorting/MergeSort.cs
using System;

class MergeSort
{
    static void Main()
    {
        int[] array = { 12, 11, 13, 5, 6, 7 };

        Console.WriteLine("Original array:");
        PrintArray(array);

        // Perform merge sort
        MergeSortAlgorithm(array, 0, array.Length - 1);

        Console.WriteLine("\nSorted array:");
        PrintArray(array);
    }

    static void MergeSortAlgorithm(int[] array, int left, int right)
    {
        if (left < right)
        {
            int middle = (left + right) / 2;

            // Recursively sort the two halves
            MergeSortAlgorithm(array, left, middle);
            MergeSortAlgorithm(array, middle + 1, right);

            // Merge the sorted halves
            Merge(array, left, middle, right);
        }
    }

    static void Merge(int[] array, int left, int middle, int right)
    {
        int n1 = middle - left + 1;
        int n2 = right - middle;

        // Create temporary arrays
        int[] leftArray = new int[n1];
        int[] rightArray = new int[n2];

        // Copy data to temporary arrays leftArray[] and rightArray[]
        for (int i = 0; i < n1; ++i)
            leftArray[i] = array[left + i];
        for (int j = 0; j < n2; ++j)
            rightArray[j] = array[middle + 1 + j];

        // Merge the temporary arrays back into arr[left..right]
        int k = left;
        int leftIndex = 0, rightIndex = 0;

        while (leftIndex < n1 && rightIndex < n2)
        {
            if (leftArray[leftIndex] <= rightArray[rightIndex])
            {
                array[k] = leftArray[leftIndex];
                leftIndex++;
            }
            else
            {
                array[k] = rightArray[rightIndex];
                rightIndex++;
            }
            k++;
        }

        // Copy the remaining elements of leftArray[], if there are any
        while (leftIndex < n1)
        {
            array[k] = leftArray[leftIndex];
            leftIndex++;
            k++;
        }

        // Copy the remaining elements of rightArray[], if there are any
        while (rightIndex < n2)
        {
            array[k] = rightArray[rightIndex];
            rightIndex++;
            k++;
        }
    }

    static void PrintArray(int[] array)
    {
        foreach (int num in array)
        {
            Console.Write(num + " ");
        }
        Console.WriteLine();
    }
}
